uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/rules/DevicePinSetupRule.java

```diff
@@ -0,0 +1,103 @@
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.client.ui.automation.rules;
+
+import android.app.KeyguardManager;
+import android.content.Context;
+import android.os.Build;
+import android.provider.Settings;
+
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.platform.app.InstrumentationRegistry;
+import androidx.test.uiautomator.UiDevice;
+import androidx.test.uiautomator.UiObject;
+import androidx.test.uiautomator.UiObjectNotFoundException;
+
+import com.microsoft.identity.client.ui.automation.utils.UiAutomatorUtils;
+
+import org.junit.Assert;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import static com.microsoft.identity.client.ui.automation.utils.CommonUtils.launchIntent;
+
+/**
+ * A test rule that allows you to setup PIN for the device if the pin is not setup and
+ * if it is already setup then it will keep the same PIN.
+ */
+public class DevicePinSetupRule implements TestRule {
+
+    static final String PASSWORD = "1234";
+
+    @Override
+    public Statement apply(final Statement base, final Description description) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                if (!isDeviceSecured()) {
+                    setLock();
+                }
+                base.evaluate();
+            }
+        };
+    }
+
+    private boolean isDeviceSecured() {
+        final Context context = ApplicationProvider.getApplicationContext();
+        final KeyguardManager keyguardManager =
+                (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            return keyguardManager.isDeviceSecure();
+        }
+        return keyguardManager.isKeyguardSecure();
+    }
+
+    private void setLock() throws UiObjectNotFoundException {
+
+        final UiDevice device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation());
+        launchIntent(Settings.ACTION_SECURITY_SETTINGS);
+
+        final UiObject screenLock = UiAutomatorUtils.obtainUiObjectWithText("Screen lock");
+        Assert.assertTrue(screenLock.exists());
+        screenLock.click();
+        final UiObject pinButton = UiAutomatorUtils.obtainUiObjectWithExactText("PIN");
+        Assert.assertTrue(pinButton.exists());
+        pinButton.click();
+        final UiObject noButton = UiAutomatorUtils.obtainUiObjectWithExactText("NO");
+        Assert.assertTrue(noButton.exists());
```

Comment:
I did not get NO button on my device and therefore test is failing....for me it goes straight to entering the PIN.

By the way, I think this entire method needs to move to the `ISettings` class and change logic based on device

Reply:
Why are we clicking on the NO button here?

Reply:
In India, I think we are getting it no Button.

Reply:
Okay let's remove the no button click....and also let's move this entire logic to `ISettings` so we can change resource ids etc...based on API level and device

Reply:
For API 28 I am getting no Button, but for API 29 I am not getting any NO Button.

Reply:
Moved the entire logic to ISettings. and it supports API 28 and API 29. and I didn't written logic for the Samsung device.

Reply:
Interesting....I did not get NO button for either API 28 or 29...so yeah probably a difference in behavior based on location....
For samsung device, you can skip it for now....I will add that logic in later